language/move-vm/types/src/natives/function.rs

```diff
@@ -36,15 +33,14 @@ pub use move_core_types::vm_status::StatusCode;
 /// Errors (typically user errors and aborts) that are logically part of the function execution
 /// must be expressed in a `NativeResult` with a cost and a VMStatus.
 pub struct NativeResult {
-    /// The cost for running that function, whether successfully or not.
-    pub cost: InternalGasUnits<GasCarrier>,
     /// Result of execution. This is either the return values or the error to report.
+    pub cost: u64,
     pub result: Result<SmallVec<[Value; 1]>, u64>,
 }
 
 impl NativeResult {
     /// Return values of a successful execution.
-    pub fn ok(cost: InternalGasUnits<GasCarrier>, values: SmallVec<[Value; 1]>) -> Self {
+    pub fn ok(cost: u64, values: SmallVec<[Value; 1]>) -> Self {
```

Comment:
Why the removal of `InternalGasUnits`?

Reply:
Previously we've been doing a lot of gas calculations inside the VM, so there was a need to not accidentally mix different gas values up. However with the new system, it is up to the GasMeter implementations to perform the actual bookkeeping, and in this particular case, the VM is simply passing whatever value native functions returns to the GasMeter. I do find the GasAlgebra monad unnecessarily complicated for our current needs so I just decided to use a plain integer instead.

Reply:
Ah, I see. 

I still have some worry with accidentally using unchecked arithmetic or something. And it might be beneficial to use the InternalGasUnits for that. But not sure it is worth the pain overall

Reply:
Thinking a bit more on this:
Not using the GasUnits system means you are going to have to manually re-check all the arithmetic for over/underflow (and any div-by-zero, though this is less likely). 
This is definitely an attack vector to consider, especially around things like hash/string/bcs/vector etc where you might be dealing with rather large sizes.

I think we either need to:
- Switch back to InternalGasUnits
- Or have some other mechanism for doing this simply inside the native functions
While it might not entirely be possible, I think it is a bit of a safety regression when compared to what was there before

Reply:
Turns out InternalGasUnits and the GasAlgebras in general are not checked. Wild